cluster-autoscaler/cloudprovider/azure/azure_client.go

```diff
@@ -138,6 +145,143 @@ func (az *azDeploymentsClient) Delete(ctx context.Context, resourceGroupName, de
+type AgentPoolsClient interface {
+	Get(ctx context.Context,
+		resourceGroupName, resourceName, agentPoolName string,
+		options *armcontainerservice.AgentPoolsClientGetOptions) (
+		armcontainerservice.AgentPoolsClientGetResponse, error)
+	BeginCreateOrUpdate(
+		ctx context.Context,
+		resourceGroupName, resourceName, agentPoolName string,
+		parameters armcontainerservice.AgentPool,
+		options *armcontainerservice.AgentPoolsClientBeginCreateOrUpdateOptions) (
+		*runtime.Poller[armcontainerservice.AgentPoolsClientCreateOrUpdateResponse], error)
+	BeginDeleteMachines(
+		ctx context.Context,
+		resourceGroupName, resourceName, agentPoolName string,
+		machines armcontainerservice.AgentPoolDeleteMachinesParameter,
+		options *armcontainerservice.AgentPoolsClientBeginDeleteMachinesOptions) (
+		*runtime.Poller[armcontainerservice.AgentPoolsClientDeleteMachinesResponse], error)
+}
+
+const userAgent = "AKS-autoscaler"
+
+func getAgentpoolClientCredentials(cfg *Config) (azcore.TokenCredential, error) {
+	var cred azcore.TokenCredential
+	var err error
+	if cfg.AuthMethod == authMethodCLI {
+		cred, err = azidentity.NewAzureCLICredential(&azidentity.AzureCLICredentialOptions{
+			TenantID: cfg.TenantID})
+		if err != nil {
+			klog.Errorf("NewAzureCLICredential failed: %v", err)
+			return nil, err
+		}
+	} else if cfg.AuthMethod == "" || cfg.AuthMethod == authMethodPrincipal {
+		cred, err = azidentity.NewClientSecretCredential(cfg.TenantID, cfg.AADClientID, cfg.AADClientSecret, nil)
+		if err != nil {
+			klog.Errorf("NewClientSecretCredential failed: %v", err)
+			return nil, err
+		}
+	} else {
+		return nil, fmt.Errorf("unsupported authorization method: %s", cfg.AuthMethod)
+	}
+	return cred, nil
+}
+
+func getAgentpoolClientRetryOptions(cfg *Config) azurecore_policy.RetryOptions {
+	if cfg.AuthMethod == authMethodCLI {
+		return azurecore_policy.RetryOptions{
+			MaxRetries: -1, // no retry when using CLI auth for UT
+		}
+	}
+	return azurecore_policy.RetryOptions{
+		MaxRetries:    5,
+		RetryDelay:    1 * time.Second,
+		MaxRetryDelay: 2 * time.Second,
+		TryTimeout:    10 * time.Second,
+	}
+}
+
+func newAgentpoolClientWithEndpoint(baseURL, subscriptionID string, retryOptions azurecore_policy.RetryOptions) (AgentPoolsClient, error) {
+	// for AKS managed CAS, we will use the ARMBaseURL to create a fake agent pool client
+	// so that the request will be sent to node provioner endpoint intead of the public ARM endpoint
+	agentPoolsClient, err := armcontainerservice.NewAgentPoolsClient(subscriptionID, nil,
+		&policy.ClientOptions{
+			ClientOptions: azurecore_policy.ClientOptions{
+				Cloud: cloud.Configuration{
+					Services: map[cloud.ServiceName]cloud.ServiceConfiguration{
+						cloud.ResourceManager: {
+							Endpoint: baseURL,
+							Audience: "UNKNOWN",
+						},
+					},
+				},
+				Telemetry: azurecore_policy.TelemetryOptions{
+					ApplicationID: userAgent,
+				},
+				Retry: retryOptions,
+			},
+		})
```

Comment:
I love to see usage of the track 2 sdk! I would recommend some shared middleware the project karpenter, fleet and some other projects consume via https://github.com/Azure/azure-sdk-for-go-extensions

There are a couple of benefits. 

1. Armbalancer will load balance the requests on the client side to prevent arm instance level throttling
2. There is a bug in go's tcp implementation for http2(which track 2 will use) where the connections are not effectively closed and we will get tcp timeout errors where the connection isn't closed for 17 minutes until it finally fails. Stephane patched this [here](https://github.com/Azure/azure-sdk-for-go-extensions/pull/29) 

We could keep going but you get the point, we want to have this shared middleware used for track 2 implementations so that you don't see the same bugs and we can share all of our throttling lessons. We maintain one core library for these arm middleware and all share the benefits. 

https://github.com/Azure/azure-sdk-for-go-extensions/blob/main/pkg/middleware/arm_client_opts.go#L25 are the arm client options i would reccomend as the base then you can override things that dont work for you.